Add reproduction case for #2629

rgrinberg opened the pull request:
The test demonstrates that it's an issue with how we parse the paths in the .install file. We can't just assume that an absolute path isn't in the build directory.

I see two solutions:

1) We just allow external paths as sources for installed things.

2) We fix the parsing to attempt to recognize external paths as build paths.

I like the 2) even without fixing this bug as I'm sure this may cause issues in other parts of the code.

rgrinberg committed:
Add reproduction case for #2629

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
I suppose both make sense. Regarding 2), I'm just concerned about parsing external paths correctly. i.e. if the build dir is `/tmp/build` and `/tmp/x` is a symlink for `/tmp/a/b/c`, then all these paths should be recognised as the build directory:

- `/tmp/build`
- `/tmp/x/../../build`
- `/tmp/a/../x/../c/../../build`
- ...

aalekseyev commented:
I think in this particular case dune is writing the .install file so a simple stripping of prefix is fine because we'll be able to read back what we wrote.

I think the logic should be specific to parsing the .install file though because of what @diml brought up.

rgrinberg commented:
> I think the logic should be specific to parsing the .install file though because of what @diml brought up.

I have a feeling that this is just choosing a different evil, and I'm not sure if it's the lesser one. Basically anywhere where we read external paths is something that can be broken by `--build-dir`.

IMO, what would be better is just having a simplified prefix check for build paths with a note somewhere that that symlinks into the build path aren't supported by dune.

diml commented:
symlinks and also all kind of non-normalised paths. In particular, we can't get rid of ".." unless we follow symlinks.

At the same time, it's likely that all paths pointing to the build directory are indeed generated by Dune. So if we assume that the user sets the build directory string once and for all, a simple string-prefix check should indeed be a good enough approximation in practice.

rgrinberg committed:
Do not rely on build_dir in ocaml_config

Parse the the standard library path using as an external path. This
guarantees that it will work regardless of build_dir.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Parse external paths in the build dir

We only do a simple prefix check to interpret such paths as in general,
we have to resolve symlinks and other relative paths for this to work.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
It looks like 2) is breaking dune cache somehow. I wonder if addressing the issue mentioned in https://github.com/ocaml/dune/pull/2639#pullrequestreview-288595010 would fix this issue.

rgrinberg commented on test/blackbox-tests/test-cases/dune-cache/run.t:
```diff
@@ -11,20 +11,23 @@
   > \_o< COIN
   > EOF
   $ env DUNE_CACHE=1 DUNE_CACHE_EXIT_NO_CLIENT=1 XDG_RUNTIME_DIR=$PWD/.xdg-runtime XDG_CACHE_HOME=$PWD/.xdg-cache dune build target
+  Error: shutdown: : Socket is not connected
```
@mefyl any idea what could be causing this?

diml commented:
Zooming out this discussion, all this feels a bit fishy. `dune install` is basically parsing a `.install` file and emitting a bunch of file copies using the filenames in this file. We should be able to do that without even knowing what the build directory is. So I propose another approach:

1. we check whether there is a place in the code where we care about the paths in `Install.Entry.t` being `Path.Build.t`
2.1 if no, then we just make them plain strings and the problem goes away entirely
2.2 if yes, we parametrise `Install.Entry.t` by the type of paths and make `load_install_file` return a `string Entry.t list`

rgrinberg committed:
Revert "Parse external paths in the build dir"

This reverts commit 3dab9b7d522676cbd26a4abb3bda7630e277e2df.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Load install files with Path.t entries

This works regardless of how --build-dir was set

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove mysterious type parameter

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Factor the code to read install files

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
>  if yes, we parametrise Install.Entry.t by the type of paths and make load_install_file return a string Entry.t list

I opted for this. Instead of `string`, I instantiated the type parameter to be `Path.t`. I'm also considering this patch for 1.11.4. The `--build-dir` feature is a bit useless if the .install files it generates are not usable by dune.

diml commented:
Including it in 1.11.4 seems good. I'm reviewing it now!
